src/main/java/org/zalando/nakadi/service/kpi/NakadiPublisher.java

```diff
@@ -0,0 +1,127 @@
+import org.zalando.nakadi.service.EventPublisher;
+import org.zalando.nakadi.util.FeatureToggleService;
+import org.zalando.nakadi.util.UUIDGenerator;
+
+import java.time.Instant;
+import java.util.Map;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+@Component
+public final class NakadiPublisher {
+
+    private final static Logger LOG = LoggerFactory.getLogger(NakadiPublisher.class);
+    private final EventPublisher eventPublisher;
+    private final Map<String, BlockingQueue<JSONObject>> eventTypeEvents;
+    private final ExecutorService executorService;
+    private final UUIDGenerator uuidGenerator;
+    private final FeatureToggleService featureToggleService;
+
+    private final long batchCollectionTimeout;
+    private final int batchSize;
+    private final long pollTimeout;
+    private final int eventsQueueSize;
+    private final long scheduleTimeout;
+
+    @Autowired
+    public NakadiPublisher(final EventPublisher eventPublisher,
+                           final UUIDGenerator uuidGenerator,
+                           final FeatureToggleService featureToggleService,
+                           @Value("${nakadi.kpi.batch-collection-timeout:1000}") final long batchCollectionTimeout,
+                           @Value("${nakadi.kpi.batch-size:3}") final int batchSize,
+                           @Value("${nakadi.kpi.workers:1}") final int workers,
+                           @Value("${nakadi.kpi.poll-timeout:100}") final long pollTimeout,
+                           @Value("${nakadi.kpi.events-queue-size:100}") final int eventsQueueSize,
+                           @Value("${nakadi.kpi.schedule-timeout:5000}") final long scheduleTimeout) {
+        this.eventPublisher = eventPublisher;
+        this.uuidGenerator = uuidGenerator;
+        this.featureToggleService = featureToggleService;
+        this.batchCollectionTimeout = batchCollectionTimeout;
+        this.batchSize = batchSize;
+        this.pollTimeout = pollTimeout;
+        this.eventsQueueSize = eventsQueueSize;
+        this.scheduleTimeout = scheduleTimeout;
+        this.eventTypeEvents = new ConcurrentHashMap<>();
+        this.executorService = Executors.newFixedThreadPool(workers);
+        this.executorService.submit(() -> scheduleEventsBatchesSend());
+    }
+
+    private void scheduleEventsBatchesSend() {
+        LOG.trace("Schedule events collection on {}", Thread.currentThread().getName());
+        if (featureToggleService.isFeatureEnabled(FeatureToggleService.Feature.KPI_COLLECTION)) {
+            for (final String etName : eventTypeEvents.keySet()) {
+                executorService.submit(() -> sendEventBatch(etName));
+            }
+        }
+        try {
+            Thread.sleep(scheduleTimeout);
```

Comment:
You can use `ScheduledExecutorService` instead and schedule `scheduleEventsBatchesSend` to be run every X seconds. And you will avoid writing your own scheduler.